File: src/servers/team_pink/worker_server.py
#!/usr/bin/env python3
"""
Python Worker Server (Process F) - Team Pink
Demonstrates cross-language gRPC integration
"""

import sys
import json
import grpc
import time
import csv
import os
from concurrent import futures
from pathlib import Path

# Add proto path
proto_path = str(Path(__file__).parent.parent.parent.parent / 'proto')
if proto_path not in sys.path:
    sys.path.insert(0, proto_path)

import fire_query_pb2
import fire_query_pb2_grpc


class WorkerServiceImpl(fire_query_pb2_grpc.FireQueryServiceServicer):
    def __init__(self, config):
        self.config = config
        self.process_id = config['process_id']
        self.team = config['team']
        self.data_path = config['data_path']
        self.owned_dates = config['data_partitioning']['owned_dates']
        self.chunk_size = config['chunk_config']['default_chunk_size']
        self.pending_requests = 0
        self.completed_requests = 0

        print(f"Worker Process {self.process_id} (Team {self.team}) starting...")
        print(f"Listening on {config['listen_host']}:{config['listen_port']}")
        print(f"Data partition: {' '.join(self.owned_dates)}")

    def DelegateQuery(self, request, context):
        """Handle delegated query from team leader"""
        print(f"\n[Worker {self.process_id}] Received delegation {request.request_id} from {request.delegating_process}")

        self.pending_requests += 1

        # Deserialize original query
        original_query = fire_query_pb2.QueryRequest()
        original_query.ParseFromString(request.original_query)

        # Determine dates to process
        dates_to_process = self._select_dates_to_process(original_query)
        print(f"  [Worker {self.process_id}] Processing {len(dates_to_process)} dates")

        if not dates_to_process:
            print(f"  [Worker {self.process_id}] No matching dates in partition")
            self.pending_requests -= 1
            self.completed_requests += 1
            return

        # Load and process data
        start_time = time.time()
        records = self._load_data(
            dates_to_process,
            original_query.pollutant_type,
            original_query.latitude_min,
            original_query.latitude_max,
            original_query.longitude_min,
            original_query.longitude_max,
            original_query.max_records
        )
        duration = (time.time() - start_time) * 1000  # Convert to ms

        print(f"  [Worker {self.process_id}] Loaded {len(records)} records in {duration:.0f}ms")

        # Send records in chunks
        chunk_count = 0
        for i in range(0, len(records), self.chunk_size):
            chunk_end = min(i + self.chunk_size, len(records))
            chunk_records = records[i:chunk_end]

            response = fire_query_pb2.DelegationResponse()
            response.request_id = request.request_id
            response.chunk_number = chunk_count
            response.is_final = False
            response.responding_process = self.process_id

            # Add records to response
            for record_data in chunk_records:
                record = response.records.add()
                self._populate_fire_record(record, record_data)

            yield response

            print(f"  [Worker {self.process_id}] Sent chunk {chunk_count} with {len(chunk_records)} records")
            chunk_count += 1

            # Simulate processing time
            time.sleep(0.05)

        print(f"[Worker {self.process_id}] Delegation {request.request_id} complete. Sent {chunk_count} chunks")

        self.pending_requests -= 1
        self.completed_requests += 1

    def HealthCheck(self, request, context):
        """Health check endpoint"""
        response = fire_query_pb2.HealthResponse()
        response.responding_process = self.process_id
        response.is_healthy = True
        response.pending_requests = self.pending_requests
        response.active_workers = 1
        return response

    def CancelQuery(self, request, context):
        """Handle query cancellation"""
        print(f"[Worker {self.process_id}] Cancel request for {request.request_id}")
        response = fire_query_pb2.CancelResponse()
        response.request_id = request.request_id
        response.cancelled = True
        return response

    def QueryFire(self, request, context):
        """Not implemented for workers"""
        context.set_code(grpc.StatusCode.UNIMPLEMENTED)
        context.set_details("Workers don't accept direct queries")
        return fire_query_pb2.QueryResponse()

    def _select_dates_to_process(self, query):
        """Select dates that match query range and are owned by this worker"""
        result = []
        for date in self.owned_dates:
            if query.date_start <= date <= query.date_end:
                result.append(date)
        return result

    def _load_data(self, dates, pollutant_filter, lat_min, lat_max, lon_min, lon_max, max_records):
        """Load fire data from CSV files"""
        results = []

        for date in dates:
            date_dir = os.path.join(self.data_path, date)
            if not os.path.exists(date_dir):
                print(f"Warning: Date directory not found: {date_dir}")
                continue

            # Load all CSV files for this date
            for csv_file in Path(date_dir).glob('*.csv'):
                records = self._load_csv(
                    str(csv_file),
                    pollutant_filter,
                    lat_min, lat_max,
                    lon_min, lon_max,
                    max_records - len(results) if max_records > 0 else -1
                )
                results.extend(records)

                if max_records > 0 and len(results) >= max_records:
                    return results[:max_records]

        return results

    def _load_csv(self, csv_path, pollutant_filter, lat_min, lat_max, lon_min, lon_max, max_records):
        """Load and parse a single CSV file"""
        results = []

        try:
            with open(csv_path, 'r') as f:
                reader = csv.reader(f, quotechar='"')
                for row in reader:
                    if max_records > 0 and len(results) >= max_records:
                        break

                    if len(row) < 13:
                        continue

                    try:
                        lat = float(row[0])
                        lon = float(row[1])
                        pollutant = row[3]

                        # Apply filters
                        if pollutant_filter and pollutant != pollutant_filter:
                            continue

                        if lat < lat_min or lat > lat_max:
                            continue

                        if lon < lon_min or lon > lon_max:
                            continue

                        # Parse full record
                        record = {
                            'latitude': lat,
                            'longitude': lon,
                            'timestamp': row[2],
                            'pollutant': pollutant,
                            'concentration': float(row[4]),
                            'unit': row[5],
                            'raw_concentration': float(row[6]),
                            'aqi': int(row[7]),
                            'aqi_category': int(row[8]),
                            'site_name': row[9],
                            'agency': row[10],
                            'site_id': row[11],
                            'full_site_id': row[12]
                        }

                        results.append(record)

                    except (ValueError, IndexError) as e:
                        # Skip malformed records
                        continue

        except FileNotFoundError:
            print(f"Warning: CSV file not found: {csv_path}")

        return results

    def _populate_fire_record(self, proto_record, data):
        """Convert Python dict to protobuf FireRecord"""
        proto_record.latitude = data['latitude']
        proto_record.longitude = data['longitude']
        proto_record.timestamp = data['timestamp']
        proto_record.pollutant = data['pollutant']
        proto_record.concentration = data['concentration']
        proto_record.unit = data['unit']
        proto_record.raw_concentration = data['raw_concentration']
        proto_record.aqi = data['aqi']
        proto_record.aqi_category = data['aqi_category']
        proto_record.site_name = data['site_name']
        proto_record.agency = data['agency']
        proto_record.site_id = data['site_id']
        proto_record.full_site_id = data['full_site_id']


def load_config(config_file):
    """Load configuration from JSON file"""
    with open(config_file, 'r') as f:
        return json.load(f)


def serve(config_file):
    """Start the Python worker server"""
    config = load_config(config_file)

    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    fire_query_pb2_grpc.add_FireQueryServiceServicer_to_server(
        WorkerServiceImpl(config), server
    )

    listen_addr = f"{config['listen_host']}:{config['listen_port']}"
    server.add_insecure_port(listen_addr)
    server.start()

    print(f"\n*** Python Worker server listening on {listen_addr} ***\n")

    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        print("\nShutting down Python worker server...")
        server.stop(0)


if __name__ == '__main__':
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <config_file>")
        sys.exit(1)

    serve(sys.argv[1])

File: src/servers/team_green/worker_server.cpp
#include <iostream>
#include <memory>
#include <string>
#include <algorithm>
#include <thread>
#include <chrono>

#include <grpc/grpc.h>
#include <grpcpp/server.h>
#include <grpcpp/server_builder.h>
#include <grpcpp/server_context.h>
#include <grpcpp/security/server_credentials.h>

#include "fire_query.grpc.pb.h"
#include "../../common/config.hpp"
#include "../../common/fire_data_loader.hpp"
#include "../../shmem/status_manager.hpp"
#include "../../common/metrics.hpp"

using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::ServerWriter;
using grpc::Status;

using firequery::FireQueryService;
using firequery::QueryRequest;
using firequery::DelegationRequest;
using firequery::DelegationResponse;
using firequery::HealthRequest;
using firequery::HealthResponse;
using firequery::CancelRequest;
using firequery::CancelResponse;
using firequery::FireRecord;

class WorkerServiceImpl final : public FireQueryService::Service {
public:
    WorkerServiceImpl(const ProcessConfig& config)
        : config_(config), status_mgr_(false), data_loader_(config.data_path) {

        std::cout << "Worker Process " << config_.process_id
                  << " (Team " << config_.team << ") starting..." << std::endl;
        std::cout << "Listening on " << config_.listen_host << ":" << config_.listen_port << std::endl;

        // Print data partition
        std::cout << "Data partition: ";
        for (const auto& date : config_.data_partitioning.owned_dates) {
            std::cout << date << " ";
        }
        std::cout << std::endl;

        // Initialize metrics logging for this process
        metrics::init_with_dir("logs", config_.process_id, config_.role);
    }

    Status DelegateQuery(ServerContext* context,
                        const DelegationRequest* request,
                        ServerWriter<DelegationResponse>* writer) override {

        std::cout << "\n[Worker " << config_.process_id << "] Received delegation "
                  << request->request_id() << " from " << request->delegating_process() << std::endl;

        metrics::log_event("RECEIVED_DELEGATION", request->request_id(), pending_requests_, 1, -1, -1, request->delegating_process());

        // Update status
        {
            std::lock_guard<std::mutex> lock(status_mutex_);
            pending_requests_++;
            status_mgr_.updateProcessStatus(config_.process_id, pending_requests_, 1, completed_requests_);
        }

        // Deserialize original query
        QueryRequest original_query;
        if (!original_query.ParseFromString(request->original_query())) {
            return Status(grpc::StatusCode::INVALID_ARGUMENT, "Failed to parse original query");
        }

        // Determine dates to process
        std::vector<std::string> dates_to_process = selectDatesToProcess(original_query);

        std::cout << "  [Worker " << config_.process_id << "] Processing "
                  << dates_to_process.size() << " dates" << std::endl;

        if (dates_to_process.empty()) {
            std::cout << "  [Worker " << config_.process_id << "] No matching dates in partition" << std::endl;
            {
                std::lock_guard<std::mutex> lock(status_mutex_);
                pending_requests_--;
                completed_requests_++;
                status_mgr_.updateProcessStatus(config_.process_id, pending_requests_, 1, completed_requests_);
            }
            return Status::OK;
        }

        // Load data
        auto start_time = std::chrono::high_resolution_clock::now();

        std::vector<FireDataRecord> records = data_loader_.loadData(
            dates_to_process,
            original_query.pollutant_type(),
            original_query.latitude_min(),
            original_query.latitude_max(),
            original_query.longitude_min(),
            original_query.longitude_max(),
            original_query.max_records()
        );

        auto end_time = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);

        std::cout << "  [Worker " << config_.process_id << "] Loaded " << records.size()
                  << " records in " << duration.count() << "ms" << std::endl;

        metrics::log_event("LOADED_RECORDS", request->request_id(), pending_requests_, 1, -1, records.size(), "loaded by worker");

        // Send in chunks
        int chunk_size = config_.chunk_config.default_chunk_size;
        int chunk_count = 0;

        for (size_t i = 0; i < records.size(); i += chunk_size) {
            DelegationResponse chunk_resp;
            chunk_resp.set_request_id(request->request_id());
            chunk_resp.set_chunk_number(chunk_count++);
            chunk_resp.set_is_final(false);
            chunk_resp.set_responding_process(config_.process_id);

            size_t end = std::min(i + chunk_size, records.size());
            for (size_t j = i; j < end; j++) {
                auto* rec = chunk_resp.add_records();
                convertToProto(records[j], rec);
            }

            if (!writer->Write(chunk_resp)) {
                std::cerr << "  [Worker " << config_.process_id << "] Failed to write chunk" << std::endl;
                {
                    std::lock_guard<std::mutex> lock(status_mutex_);
                    // Metrics: failed to send worker chunk upstream
                    metrics::log_event("WORKER_CHUNK_SEND_ERROR", request->request_id(), pending_requests_, 1, chunk_resp.chunk_number(), chunk_resp.records_size(), config_.process_id);
                    pending_requests_--;
                    status_mgr_.updateProcessStatus(config_.process_id, pending_requests_, 1, completed_requests_);
                }
                return Status::CANCELLED;
            }

            // Metrics: worker chunk sent (only after successful write)
            metrics::log_event("WORKER_CHUNK_SENT", request->request_id(), pending_requests_, 1, chunk_resp.chunk_number(), chunk_resp.records_size(), config_.process_id);

            std::cout << "  [Worker " << config_.process_id << "] Sent chunk " << chunk_count - 1
                      << " with " << chunk_resp.records_size() << " records" << std::endl;

            // Simulate some processing time for realistic demonstration
            std::this_thread::sleep_for(std::chrono::milliseconds(50));
        }

        std::cout << "[Worker " << config_.process_id << "] Delegation "
                  << request->request_id() << " complete. Sent " << chunk_count << " chunks" << std::endl;

        // Update status
        {
            std::lock_guard<std::mutex> lock(status_mutex_);
            pending_requests_--;
            completed_requests_++;
            status_mgr_.updateProcessStatus(config_.process_id, pending_requests_, 1, completed_requests_);
        }

        return Status::OK;
    }

    Status HealthCheck(ServerContext* context,
                      const HealthRequest* request,
                      HealthResponse* response) override {
        response->set_responding_process(config_.process_id);
        response->set_is_healthy(true);
        response->set_pending_requests(pending_requests_);
        response->set_active_workers(1);
        return Status::OK;
    }

    // Not used by workers
    Status QueryFire(ServerContext* context,
                    const QueryRequest* request,
                    ServerWriter<firequery::QueryResponse>* writer) override {
        return Status(grpc::StatusCode::UNIMPLEMENTED, "Workers don't accept direct queries");
    }

    Status CancelQuery(ServerContext* context,
                      const CancelRequest* request,
                      CancelResponse* response) override {
        std::cout << "[Worker " << config_.process_id << "] Cancel request for "
                  << request->request_id() << std::endl;
        response->set_request_id(request->request_id());
        response->set_cancelled(true);
        return Status::OK;
    }

private:
    ProcessConfig config_;
    StatusManager status_mgr_;
    FireDataLoader data_loader_;
    int pending_requests_ = 0;
    int completed_requests_ = 0;
    std::mutex status_mutex_;

    std::vector<std::string> selectDatesToProcess(const QueryRequest& query) {
        std::vector<std::string> result;

        for (const auto& date : config_.data_partitioning.owned_dates) {
            if (date >= query.date_start() && date <= query.date_end()) {
                result.push_back(date);
            }
        }

        return result;
    }

    void convertToProto(const FireDataRecord& src, FireRecord* dest) {
        dest->set_latitude(src.latitude);
        dest->set_longitude(src.longitude);
        dest->set_timestamp(src.timestamp);
        dest->set_pollutant(src.pollutant);
        dest->set_concentration(src.concentration);
        dest->set_unit(src.unit);
        dest->set_raw_concentration(src.raw_concentration);
        dest->set_aqi(src.aqi);
        dest->set_aqi_category(src.aqi_category);
        dest->set_site_name(src.site_name);
        dest->set_agency(src.agency);
        dest->set_site_id(src.site_id);
        dest->set_full_site_id(src.full_site_id);
    }
};

void RunWorkerServer(const std::string& config_file) {
    ProcessConfig config = ConfigParser::loadConfig(config_file);

    std::string server_address = config.listen_host + ":" + std::to_string(config.listen_port);
    WorkerServiceImpl service(config);

    ServerBuilder builder;
    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
    builder.RegisterService(&service);

    std::unique_ptr<Server> server(builder.BuildAndStart());
    std::cout << "\n*** Worker server listening on " << server_address << " ***\n" << std::endl;

    server->Wait();
}

int main(int argc, char** argv) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <config_file>" << std::endl;
        return 1;
    }

    try {
        RunWorkerServer(argv[1]);
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        metrics::shutdown();
        return 1;
    }

    metrics::shutdown();
    return 0;
}
